### ISSUE TITLE
Leader election threads exit on error, and don't restart

### ISSUE THREAD

**Issue by DaveLangridge** [projectcalico/felix#543]
neutron-ha code.

E.g. etcd exception causes:

2015-05-14 15:12:29.633 12905 ERROR calico.election [-] Election greenlet exiting

but the thread doesn't restart. Can result in no leader.

**Comment by Lukasa** [comment]
I have a candidate fix in #545.

**Comment by fasaxc** [comment]
@Lukasa should we take down the process if our leader election or PR threads actually do die?  I'd guess that Neutron will restart the child worker.

**Comment by Lukasa** [comment]
No, I don't think so: I think we should restart them. They should default to 'safe' (no resync), but if it's possible for us to not take out our process we should do that.

**Comment by fasaxc** [comment]
I'm wary of trying too hard (call it the Erlang philosophy).  Neutron appears to supervise and restart children.  Dying and being restarted covers all sorts of cases that we can't easily spot or take care of ourselves: memory leaks; memory corruption; bugs in our code that only manifest after we've been running a while.  It doesn't help if we get stuck in a restart loop at start of day, but nor does restarting the thread without tearing down the process...

**Comment by plwhite** [comment]
In almost every case now we catch exceptions and avoid exiting the greenlet. However, there are some examples where we might still hit problems. For example, suppose we get some random exception from the initial read? I think it is reasonable for us to have a very broad except in _run that prevents the greenlet ever exiting, no matter what, so I'll do a PR with that change in it.

**Comment by fasaxc** [comment]
Hmm, I'm still not convinced, we don't want to catch errors caused by memory/FD leaks, for example.  Isn't it better to just let the child die and get restarted if we hit an unexpected exception?  Neutron will do that for us if we're in an API worker, otherwise the init daemon should do so.

**Comment by lxpollitt** [comment]
+1 from me for @fasaxc's point of view that die is a useful code path to have and one we should be aiming to exercise frequently enough that we know it is robust (i.e. die more frequently than you might strictly have to in order to make sure it is robust for the times when there really is no choice but to die).  The balance is obviously that we still only want to go through the die path fairly infrequently (assuming there is some sort of down time or during the die process).

**Comment by Lukasa** [comment]
This is an ongoing discussion, but it's not currently a bug that we're desperately worried about.

**Comment by neiljerram** [comment]
I've had cause to revisit this issue.  A customer has reported that if the etcd proxy on a controller node is stopped, and started again after first waiting for long enough for 'neutron agent-list' to show the Calico agents as all down, the Calico agents are never reported as up again, until there is an attempt to launch a new VM.  (And that attempt to launch a new VM will fail, because of all the agents being down.)

That agent status problem is related to this issue: the underlying problem is that the leader election thread has died and the Elector reports that it is no longer the master; therefore `_status_reporting_thread` does not run an EtcdWatcher to watch for agent status updates.

One possible solution (per @fasaxc and @lxpollitt above) is to do `sys.exit` at the point where the elector thread is about to exit.  If I try that, I observe that:

- the Neutron API becomes intermittently unavailable, during the period when the etcd proxy is down; for example, a periodic `neutron agent-list` command sometimes works normally and sometimes reports

    Unable to establish connection to http://172.31.48.14:9696/v2.0/agents.json

- when the etcd proxy is restarted, the Calico agents are quickly (within 10 seconds) reported as up again.

In other words, it fixes the problem at the cost of (or with the feature of?) tying the whole Neutron server's fate to the etcd connection.

Thoughts?  I think that a working etcd cluster is fundamental to Calico/OpenStack operation, so it is no worse for the Neutron API to be down than for the etcd API to be unavailable (during a window where an admin has deliberately stopped etcd for some reason); hence I think this is a decent solution.

**Comment by bertjwregeer** [comment]
@neiljerram 

> Unable to establish connection to http://172.31.48.14:9696/v2.0/agents.json

When this happens, is the entire service down? As in, will HAProxy or other HA solutions see that the node is down and remove it from their pool so that if for example we have three neutron-server's running and only a single node out of the three is affected, then requests will correctly flow to the other HA nodes?

Or does the connection succeed but never send data/timeout which would cause issues even if running in HA capacity?

**Comment by neiljerram** [comment]
@bertjwregeer I haven't tested yet, but I would expect that only the Neutron server is affected on the node where etcd has been taken down.  Hence, if you have an HA setup for Neutron, the other servers should continue to be available, and HAProxy should direct requests to them.

I will test this today.

I guess the implication of your question is that it would be an acceptable fix, if the Neutron service as a whole is still available through other servers.  Is that correct?

**Comment by neiljerram** [comment]
@bertjwregeer With two Neutron servers, and HAProxy, and the proposed fix...

- If one server's etcd proxy is stopped (and so that Neutron server goes into a slow restart loop), it is still possible to observe `Unable to establish connection to http://172.31.48.14:9696/v2.0/agents.json` sometimes, for the Neutron service overall.  I believe this is because that Neutron server is going up and down, and it is possible for HAProxy to think it is up and direct a query to it, just as it goes down again :-)

- However, there is a trivial administrative workaround for this: simply do `service neutron-server stop` as well as `service etcd stop`, when you have reason to take etcd out of service.  Then HAProxy will reliably direct queries to the other Neutron server, and the Neutron service remains reliably available.

Thoughts?

**Comment by bertjwregeer** [comment]
All of the issues mentioned here has the trivial administrative work-around of simply restarting both `neutron-server` and `etcd`, however that doesn't necessarily solve the underlying problem. ;-)

I wonder if there is a way to stop `neutron-server` from going up and down so that HAProxy simply sees it as down until the `etcd` worker is back online. Is there a way to stop the HTTP server part of it starting until another thread signals it should start?

**Comment by fasaxc** [comment]
@neiljerram I think there are two bugs here.

1. The elector thread is not handling the exception that it gets when etcd is down. That's a completely expected case so we should be handling it correctly and continue to retry.  The thread should not die in that case.

1. When the elector thread dies, the process stays up in a broken state, which is the worst of all worlds (appears to be up but actually down).  If we hit something unexpected that kills the elector, we _should_ kill neutron so that the whole process gets restarted.

**Comment by neiljerram** [comment]
Thanks for your last comments, @bertjwregeer and @fasaxc.  @bertjwregeer, I don't think it's possible to make `neutron-server` behave as you suggest - which is a shame as I agree that would be the best behavior; then a sufficiently intelligent load balancer could avoid it, and the Neutron server as a whole would still be reliable, in an HA setup.

So, we have a choice about whether `neutron-server` shares fate with its local etcd proxy.

1. It does: then `neutron-server` keeps trying to restart, and dying again.  Intermittent API availability can only confuse the load balancer.  Neutron API requests may be directed to us but then fail, regardless of whether they involve Calico.

2. It doesn't: then `neutron-server` API stays available all the time, but that server will fail requests that involve Calico.  Requests that don't involve Calico should reliably succeed, whichever `neutron-server` they are directed to.

It seems clear that (2) is better, so I'm going to work on @fasaxc's suggestions, which will give us (2) in the scenario where the local etcd proxy stops for any reason.

**Comment by bertjwregeer** [comment]
@neiljerram Is there a way to add an endpoint that the HAProxy could hit to check for liveness that would allow HAProxy to mark it as down if Calico is not working?

This would solve the problem, in a slightly different way, but if documented could be a way to allow for HA without sending traffic to a neutron-server that can't handle the request.

**Comment by neiljerram** [comment]
I don't think so, because the HAProxy is in general on another machine, and the etcd proxy listens (when it is up) on 127.0.0.1.

**Comment by bertjwregeer** [comment]
@neiljerram I specifically meant adding an end-point to neutron-server.

For example:

`/v2.0/calico/health`

or something along those lines. I am not familiar enough with `neutron` and the extension API, but is something like that possible?

**Comment by neiljerram** [comment]
This issue has now been fixed by https://review.openstack.org/378727.
